grails-app/services/org/pih/warehouse/tableroapi/IndicatorDataService.groovy

```diff
@@ -385,4 +386,55 @@ class IndicatorDataService {
 
         return indicatorData;
     }
+
+    def getDelayedShipments(Location location) {
+        Date oneWeekAgo = LocalDate.now().minusWeeks(1).toDate()
+        Date oneMonthAgo = LocalDate.now().minusMonths(1).toDate()
+        Date twoMonthsAgo = LocalDate.now().minusMonths(2).toDate()
+
+        def results = Shipment.executeQuery("""
+            select s.shipmentType.id, s.shipmentNumber, s.name, s.id
+            from Shipment as s
+            inner join s.currentEvent as e
+            where s.destination = :location
+            and s.currentStatus in ('SHIPPED', 'PARTIALLY_RECEIVED')
+            and (
+                (s.shipmentType.id = 1 and e.eventDate < :oneMonthAgo)
+                or (s.shipmentType.id = 2 and e.eventDate < :twoMonthsAgo)
+                or (s.shipmentType.id in (3, 4) and e.eventDate < :oneWeekAgo)
+            )
+        """, [
+                'location'    : location,
+                'oneWeekAgo'  : oneWeekAgo,
+                'oneMonthAgo' : oneMonthAgo,
+                'twoMonthsAgo': twoMonthsAgo
+        ])
+
+        def numberDelayed = [
+                air            : 0,
+                sea            : 0,
+                landAndSuitcase: 0,
+        ]
+
+        results = results.collect {
```

Comment:
Would it be possible to replace this by GROUP BY and COUNT in the SQL query?

Reply:
We can't get the aggregated count for each shipment type and the shipment information in the same query. We could do a second query to get the count, but we would still need to iterate over the first result to format it.